chain/network/src/peer_manager/connection_store/mod.rs

```diff
@@ -34,51 +33,33 @@ impl Inner {
         return false;
     }
 
-    fn update(&mut self, clock: &time::Clock, state: Arc<NetworkState>) {
-        let now = clock.now();
-        let now_utc = clock.now_utc();
+    fn remove_outbound(&mut self, peer_id: &PeerId) {
+        self.outbound.retain(|c| c.peer_info.id != *peer_id);
+        if let Err(err) = self.store.set_recent_outbound_connections(&self.outbound) {
+            tracing::error!(target: "network", ?err, "Failed to save recent outbound connections");
+        }
+    }
 
+    /// Takes a list of ConnectionInfos and inserts them to the outbound store
+    fn insert_outbound(&mut self, conns: Vec<ConnectionInfo>) {
         // Start with the connections already in storage
         let mut updated_outbound: HashMap<PeerId, ConnectionInfo> =
             self.outbound.iter().map(|c| (c.peer_info.id.clone(), c.clone())).collect();
 
-        // Add information about live outbound connections
-        for c in state.tier2.load().ready.values() {
-            if c.peer_type != PeerType::Outbound {
-                continue;
-            }
-
-            // If this connection is already in storage, update its last_connected time
-            if let Some(stored) = updated_outbound.get_mut(&c.peer_info.id) {
-                stored.last_connected = now_utc;
-                continue;
-            }
-
-            // If this connection is not in storage and has lasted long enough, add it as a new entry
-            let connected_duration: time::Duration = now - c.established_time;
-            if connected_duration > STORED_CONNECTIONS_MIN_DURATION {
-                let first_connected: time::Utc = now_utc - connected_duration;
-
-                updated_outbound.insert(
-                    c.peer_info.id.clone(),
-                    ConnectionInfo {
-                        peer_info: c.peer_info.clone(),
-                        first_connected: first_connected,
-                        last_connected: now_utc,
-                    },
-                );
-            }
+        // Insert the new connections
+        for conn_info in conns {
+            updated_outbound.insert(conn_info.peer_info.id.clone(), conn_info);
```

Comment:
shouldn't you check that the new conn_info is actually newer than the existing one (i.e. that time_connected_until is actually larger than what is stored)? Also please document what will happen if UTC clock is rewinded - such case should be handled in a sane (but not necessarily useful) way. What I mean is that it might for example wipe ConnectionStore, but it shouldn't make the server suddenly drop all the connections for example.

Also, perhaps in such case we should wipe all the connections ending in the future, so that if clock is totally skewed at some point in time (like 1 year in the future), then as soon as it is synced back, it should start working (we shouldn't need to wait for 1 year for the broken cache entries to be wiped out).

Reply:
Also recovery from a clock skew should be covered by a unit test

Reply:
We only push info about active connections and active connections are always the most-recently-connected. So, the cleanest thing seems to be not worrying about timestamps at all and just pushing new conn_infos to the front of the store. We can still store timestamps for debugging purposes but they are no longer considered at all for ordering/eviction.

Reply:
oh, I see that you already did that. Thanks.